src/uct/ib/rc/base/rc_iface.c

```diff
@@ -165,14 +180,65 @@ static void uct_rc_iface_set_path_mtu(uct_rc_iface_t *iface,
     }
 }
 
-UCS_CLASS_INIT_FUNC(uct_rc_iface_t, uct_ib_iface_ops_t *ops, uct_pd_h pd,
+void uct_rc_iface_handle_fc(uct_rc_iface_t *iface, struct ibv_wc *wc,
+                            uct_ib_iface_recv_desc_t *desc)
+{
+    ucs_status_t status;
+    int16_t      cur_wnd;
+    uct_rc_iface_ops_t *rc_ops;
+    uct_rc_ep_t  *ep  = uct_rc_iface_lookup_ep(iface, wc->qp_num);
+    uct_rc_hdr_t *hdr = uct_ib_iface_recv_desc_hdr(&iface->super, desc);
+    uint8_t fc_hdr    = uct_rc_ep_get_fc_hdr(hdr->am_id);
+
+    if (fc_hdr & UCT_RC_EP_FC_FLAG_GRANT) {
+        /* Got either grant flag or special FC grant message */
+        cur_wnd = ep->fc_wnd;
+
+        /* Peer granted resources, so update wnd */
+        ep->fc_wnd = iface->config.fc_wnd_size;
+
+        /* To preserve ordering we have to dispatch all pending
+         * operations if current fc_wnd is <= 0
+         * (otherwise it will be dispatched by tx progress) */
+        if (cur_wnd <= 0) {
+            ucs_arbiter_group_schedule(&iface->tx.arbiter, &ep->arb_group);
+            ucs_arbiter_dispatch(&iface->tx.arbiter, 1,
+                                 uct_rc_ep_process_pending, NULL);
+        }
+        if  (fc_hdr == UCT_RC_EP_FC_PURE_GRANT) {
+            /* Special FC grant message can't be bundled with any other FC
+             * request. Stop processing this AM and do not invoke AM handler */
+            ucs_mpool_put_inline(desc);
+            return;
+        }
+    }
+    if (fc_hdr & UCT_RC_EP_FC_FLAG_SOFT_REQ) {
+        /* Got soft credit request. Mark ep that it needs to grant
+         * credits to the peer in outgoing AM (if any). */
+        ep->flags |= UCT_RC_EP_FC_FLAG_GRANT;
```

Comment:
No (at least in the current version). The protocol assumes that if receiver handles request and sends credit grant, it has receive resources. Sender just updates its window to the default value when receive grant message. It is not really intelligent, but simple and non-intrusive. Otherwise we would need to track all incoming requests to make fair resource distribution, etc